### PR TITLE
Sts region passthrough when using assume_role_web_identity_credentials

### LINKED ISSUE #2126: AWS IRSA - Region not being passed from created Client to STS to get JWT
Please fill out the sections below to help us address your issue

### Issue description
[IAM Roles for Service Accounts](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) has been released in AWS for Kubernetes clusters running v1.13+
In the linked article, it mentions a minimum version of the Ruby SDK required to interact natively with this service (without manually making a STS call via the command line and setting env variables) is `Ruby 2.11.345`

*The issue I am seeing* is that when I pass a region parameter to instantiate a Client as part of my code, this region is not being passed to the STS call - this error occurs in all three of the clients I have tried (IAM, S3, EC2).
I assume this is something to do with the new handling of the `assume_role_web_identity_credentials`, released by AWS as part of this.

When using our original IAM implementation with kube2IAM, the same code does not cause any errors - only when using "IAM roles for Service Accounts"

A workaround, for anyone reading pre-fix, is to set `Aws.config.update({ region: 'region' })` at the start of your code, rather than set it as part of the client, e.g `ec2 = Aws::EC2::Client.new(region: 'us-west-2')` - but this would not be a solution for large codebases this might affect.

[Configuration options available](https://github.com/aws/aws-sdk-ruby#configuration-options)

Error:
```
/usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/plugins/regional_endpoint.rb:44:in `after_initialize': missing region; use :region option or export region name to ENV['AWS_REGION'] (Aws::Errors::MissingRegionError)
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:78:in `block in after_initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:77:in `each'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:77:in `after_initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:21:in `initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-sts/client.rb:252:in `initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:99:in `new'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/assume_role_web_identity_credentials.rb:55:in `initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/credential_provider_chain.rb:97:in `new'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/credential_provider_chain.rb:97:in `assume_role_web_identity_credentials'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/credential_provider_chain.rb:12:in `block in resolve'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/credential_provider_chain.rb:11:in `each'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/credential_provider_chain.rb:11:in `resolve'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/aws-sdk-core/plugins/credentials_configuration.rb:53:in `block in <class:CredentialsConfiguration>'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:70:in `call'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:213:in `block in resolve_defaults'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:57:in `each'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:57:in `each'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:212:in `resolve_defaults'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:205:in `value_at'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:189:in `block in resolve'
        from /usr/local/lib/ruby/2.5.0/set.rb:338:in `each_key'
        from /usr/local/lib/ruby/2.5.0/set.rb:338:in `each'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:189:in `resolve'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:177:in `apply_defaults'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/configuration.rb:150:in `build!'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:62:in `build_config'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:19:in `initialize'
        from /usr/local/bundle/gems/aws-sdk-ec2-1.110.0/lib/aws-sdk-ec2/client.rb:256:in `initialize'
        from /usr/local/bundle/gems/aws-sdk-core-3.68.0/lib/seahorse/client/base.rb:99:in `new'
        from ./app.rb:45:in `<main>'
```

This PR may be a fix - but I am not sure enough of how the gems work to make that assumption
https://github.com/aws/aws-sdk-ruby/pull/2090

### Gem name ('aws-sdk', 'aws-sdk-resources' or service gems like 'aws-sdk-s3') and its version

Has been observed using the following gems - but may affect others
`aws-sdk-s3   1.48.0`
`aws-sdk-ec2 1.110.0`
`aws-sdk-iam 1.30.0`

### Version of Ruby, OS environment

Ruby 2.5.1
Running on the `ruby:2.5` Docker base image
Running inside a Kubernetes Cluster with v1.13 and an OIDC provider for IRSA

### Code snippets / steps to reproduce

Needs to be using the IRSA features as a pod in AWS EKS or another Cluster.

Causes an issue:
```
require 'aws-sdk-s3'
require 'aws-sdk-ec2'
require 'aws-sdk-iam'
require 'yaml'
​
region = 'us-east-1'
​
# Logging for Docker
if ENV['DOCKER_LOGS']
  fd = IO.sysopen("/proc/1/fd/1", "w")
else
  fd = IO.sysopen("/dev/stdout", "w")
end
​
a = IO.new(fd,"w")
a.sync = true

begin
  clientEC2 = Aws::EC2::Client.new({ region: region }) # <== Causes the issue
​
  a.puts "\n Getting Worker Nodes"
  resp = clientEC2.describe_instances({
    filters: [
      {
        name: "tag:KubernetesCluster",
        values: [
          "<kluster_name>"
        ]
      },
      {
        name: "vpc-id",
        values: [
          "<vpc_id>"
        ]
      }
    ]
  })
​
  resp.reservations.each do |r|
    r.instances.each do |i|
      a.puts i.instance_id
      a.puts i.private_dns_name
      a.puts i.security_groups[0].group_name
    end
  end
​
  a.puts "\n Getting Tags for single Worker Node"
  resp = clientEC2.describe_tags({
    filters: [
      {
        name: "resource-id",
        values: [
          "<ami_id>"
        ]
      }
    ]
  })
​
  a.puts YAML.dump resp.to_h
rescue Aws::EC2::Errors::ServiceError => e
  a.puts "#{e}"
end
```

Working scenario:
```
require 'aws-sdk-s3'
require 'aws-sdk-ec2'
require 'aws-sdk-iam'
require 'yaml'
​
region = 'us-east-1'
​
# Logging for Docker
if ENV['DOCKER_LOGS']
  fd = IO.sysopen("/proc/1/fd/1", "w")
else
  fd = IO.sysopen("/dev/stdout", "w")
end
​
a = IO.new(fd,"w")
a.sync = true

Aws.config.update({ region: region })

begin
  clientEC2 = Aws::EC2::Client.new # <== No error
​
  a.puts "\n Getting Worker Nodes"
  resp = clientEC2.describe_instances({
    filters: [
      {
        name: "tag:KubernetesCluster",
        values: [
          "<kluster_name>"
        ]
      },
      {
        name: "vpc-id",
        values: [
          "<vpc_id>"
        ]
      }
    ]
  })
​
  resp.reservations.each do |r|
    r.instances.each do |i|
      a.puts i.instance_id
      a.puts i.private_dns_name
      a.puts i.security_groups[0].group_name
    end
  end
​
  a.puts "\n Getting Tags for single Worker Node"
  resp = clientEC2.describe_tags({
    filters: [
      {
        name: "resource-id",
        values: [
          "<ami_id>"
        ]
      }
    ]
  })
​
  a.puts YAML.dump resp.to_h
rescue Aws::EC2::Errors::ServiceError => e
  a.puts "#{e}"
end
```

### PR DESCRIPTION
**Pull request by alextwoods**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/master/CONTRIBUTING.md

Thank you for your contribution!

fixes #2126  

## Overview
Allow pass-through of region parameter to STS client when using assume_role_web_identity_credentials.  Note:  we could pass through additional parameters, however, this may have unintended consequences.

### CODE REVIEW

File: gems/aws-sdk-core/spec/aws/credential_resolution_chain_spec.rb [hunk 1]
```diff
@@ -91,11 +100,17 @@ module Aws
           'AR_TOKEN'
         )
         client = ApiHelper.sample_rest_xml::Client.new(
-          profile: 'ar_plus_creds', region: 'us-east-1'
+          profile: 'ar_plus_creds', region: 'us-east-2'
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/credential_resolution_chain_spec.rb]: Is this trying to demonstrate the two regions? What about the classical us-west-2?

File: gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb [hunk 2]
```diff
@@ -117,16 +117,18 @@ def assume_role_credentials_from_config(opts = {})
...
 
-    def assume_role_web_identity_credentials_from_config(profile)
-      p = profile || @profile_name
+    def assume_role_web_identity_credentials_from_config(opts = {})
+      p = opts[:profile] || @profile_name
       if @config_enabled && @parsed_config
         entry = @parsed_config.fetch(p, {})
         if entry['web_identity_token_file'] && entry['role_arn']
-          AssumeRoleWebIdentityCredentials.new(
+          cfg = {
```
[the remark this answers is not in the document]
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb]: Yeah, wasn't sure if this is better or worse:
```ruby
          AssumeRoleWebIdentityCredentials.new(
            {
              role_arn: entry['role_arn'],
              web_identity_token_file: entry['web_identity_token_file'],
              role_session_name: entry['role_session_name']
            }.tap { |cfg| cfg[:region] = opts[:region] if opts[:region] }
          )
```
Its slightly shorter, but slightly less clear.  I ended up going with the current.  thoughts?

File: gems/aws-sdk-core/spec/aws/credential_resolution_chain_spec.rb (hunk 1 above)
Comment (alextwoods) [reply gems/aws-sdk-core/spec/aws/credential_resolution_chain_spec.rb]: Yeah - was just making sure it was passing something different.  I'll change to us-west-2 for tradition.

File: gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb]: Meh, what you have now is fine. Maybe going forward, tap makes sense when modifying at least 2 things.

**Comment by aackerman** [comment]
This is great, I appreciate the fast turn around @alextwoods 🎉

**Pull request merged**